File: packages/schemas/src/theme.ts
import { z } from "zod";
import { Id } from "./primitives.js";
import { Sentiment } from "./signal.js";
import { ThemeConfidence } from "./confidence.js";

export const ThemeClaim = z.object({
  description: z.string().min(1),
}).strict();
export type ThemeClaim = z.infer<typeof ThemeClaim>;

export const Theme = z.object({
  id: Id,
  topic: z.string().min(1),
  sentiment: Sentiment,
  signalIds: z.array(Id).min(1),
  confidence: ThemeConfidence,
  maySuggest: z.array(ThemeClaim),
  mayNotSuggest: z.array(ThemeClaim),
  limitations: z.array(z.string().min(1)),
});
export type Theme = z.infer<typeof Theme>;

File: backend/src/main/java/com/acltabontabon/kuro/domain/Theme.java
package com.acltabontabon.kuro.domain;

import java.util.List;

// mirrors packages/schemas/src/theme.ts
public record Theme(
        String id,
        String topic,
        Sentiment sentiment,
        List<String> signalIds,
        ThemeConfidence confidence,
        List<ThemeClaim> maySuggest,
        List<ThemeClaim> mayNotSuggest,
        List<String> limitations) {

    public Theme {
        signalIds = List.copyOf(signalIds);
        maySuggest = List.copyOf(maySuggest);
        mayNotSuggest = List.copyOf(mayNotSuggest);
        limitations = List.copyOf(limitations);
    }

    public record ThemeClaim(String description) {
    }
}
